### PR TITLE
Add mono test runner

### PR DESCRIPTION
**Pull request by khyperia**
Related: https://github.com/dotnet/roslyn/pull/23747

### CODE REVIEW

File: build/scripts/obtain_mono.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    local MONO_VERSION="$(get_tool_version mono)"
+    # the tar file has `mono` as the root directory
+    local MONO_PATH="${THIS_DIR}"/../../Binaries/Tools
+
+    if [[ ! -x "${MONO_PATH}/mono/bin/mono" ]]
+    then
+        echo "Downloading mono ${MONO_VERSION}"
+        mkdir -p "${MONO_PATH}"
+        curl -L https://roslyninfra.blob.core.windows.net/jenkins/mono/mono-${MONO_VERSION}.tar.gz | tar xz -C "${MONO_PATH}"
```
Comment (jasonmalinowski) [inline build/scripts/obtain_mono.sh]: Is there a reason we're having to use a (apparently custom) version of mono vs. some official .deb? I'm assuming there *is* a good reason, but wonder where we've documented that. :smile:

File: build/scripts/build_mono.sh [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+
+mkdir -p ../../Binaries/mono_build
+cd ../../Binaries/mono_build
+
+sudo apt-get install git autoconf libtool automake build-essential mono-devel gettext cmake
```
Comment (jasonmalinowski) [inline build/scripts/build_mono.sh]: Good that we're installing git on a machine that already cloned this repo to begin with. ;-)

**Comment by jasonmalinowski** [comment]
@dotnet-bot test ci please.

**Comment by jaredpar** [comment]
@jasonmalinowski  Never utter that phrase!!! It's like feeding a gremlin after midnight.

File: build/scripts/build_mono.sh (hunk 2 above)
Comment (jaredpar) [reply build/scripts/build_mono.sh]: Where would you expect such documentation to exist?

File: build/scripts/obtain_mono.sh (hunk 1 above)
Comment (jaredpar) [reply build/scripts/obtain_mono.sh]: Where would you expect such documentation to exist? The version is in Tools.props (as is the case with other external tools we depend on),

Comment (jasonmalinowski) [reply build/scripts/obtain_mono.sh]: Lacking a better idea, a comment before that curl line.

File: build/scripts/build_mono.sh (hunk 2 above)
Comment (khyperia) [reply build/scripts/build_mono.sh]: This is copied from http://www.mono-project.com/docs/compiling-mono/linux/#debian-based-distributions

> already cloned this repo

Not necessarily. One could have downloaded this repo via tarball or zip.

File: build/scripts/obtain_mono.sh (hunk 1 above)
Comment (khyperia) [reply build/scripts/obtain_mono.sh]: > some official .deb?

How would you install the .deb without root? (you can't)

If your answer is "unpack it", there is no single .deb - how do you get a list of the required .debs? (`mono-complete` depends on hundreds of packages)

If your answer is "download the package lists/dependency graph from mono's package servers and find the transitive closure", congrats, you've reimplemented a linux package manager. I didn't want to do that.

**Pull request merged**